File: app/src/main/java/com/martinb/marvelapp/ui/CharacterAdapter.java
package com.martinb.marvelapp.ui;

import android.content.Context;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.martinb.marvelapp.R;
import com.martinb.marvelapp.data.model.Result;
import com.squareup.picasso.Picasso;

import java.util.List;

public class CharacterAdapter extends RecyclerView.Adapter<CharacterAdapter.ViewHolder> {

    private List<Result> mCharacter;
    private LayoutInflater mInflater;
    private OnItemClicked onItemClickedListener;


    CharacterAdapter(Context context, List<Result> data) {
        this.mInflater = LayoutInflater.from(context);
        this.mCharacter = data;
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.hero_layout, parent, false);
        return new ViewHolder(view);
    }

    public void setOnItemClickedListener(OnItemClicked onItemClickedListener) {
        this.onItemClickedListener = onItemClickedListener;
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final Result character = mCharacter.get(position);
        holder.characterName.setText(character.getName());
        Picasso.get().load(Uri.parse(String.format("%s.%s",character.getThumbnail().getPath(),character.getThumbnail().getExtension()))).into(holder.characterImage);
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClickedListener.onItemSelected(character);
            }
        });
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return mCharacter.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView characterName;
        ImageView characterImage;

        ViewHolder(View itemView) {
            super(itemView);
            characterImage = itemView.findViewById(R.id.characterImage);
            characterName = itemView.findViewById(R.id.characterName);
        }

    }

}

File: app/src/main/java/com/martinb/marvelapp/ui/CharacterAdapter.kt
package com.martinb.marvelapp.ui

import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.martinb.marvelapp.R
import com.martinb.marvelapp.data.model.Result
import com.squareup.picasso.Picasso


class CharacterAdapter(val context: Context, var data: List<Result>,val listener: OnItemClicked) : RecyclerView.Adapter<CharacterAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.hero_layout, parent, false))
    }

    fun setNewCharactersData(newData: List<Result>){
        this.data = newData
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val character = data[position]
        holder.characterName.text = character.name
        holder.characterDescription.text = if(character.description == "") "No description avaliable!!" else character.description
        holder.bind(data[position], listener)
        Picasso.get().load(Uri.parse(String.format("%s/standard_fantastic.%s", character.thumbnail?.path,character.thumbnail?.extension))).into(holder.characterImage)
    }

    override fun getItemCount(): Int = data.size


    class ViewHolder (view: View) : RecyclerView.ViewHolder(view) {
        val characterImage = view.findViewById<ImageView>(R.id.characterImage)
        val characterName = view.findViewById<TextView>(R.id.characterName)
        val characterDescription = view.findViewById<TextView>(R.id.characterDescription)

        fun bind(character: Result,listener: OnItemClicked){
            itemView.setOnClickListener {
                listener.onItemClicked(character)
            }
        }
    }
}
